StringIndexOutOfBoundsException in Activity_RingtonePicker

This crash report was received from Google Play.

Device: Samsung Galaxy A8 (2018)
Android version: 9 (SDK 28)

```java
java.lang.RuntimeException: 
  at android.app.ActivityThread.deliverResults (ActivityThread.java:4617)
  at android.app.ActivityThread.handleSendResult (ActivityThread.java:4659)
  at android.app.servertransaction.ActivityResultItem.execute (ActivityResultItem.java:49)
  at android.app.servertransaction.TransactionExecutor.executeCallbacks (TransactionExecutor.java:108)
  at android.app.servertransaction.TransactionExecutor.execute (TransactionExecutor.java:68)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1955)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loop (Looper.java:214)
  at android.app.ActivityThread.main (ActivityThread.java:7078)
  at java.lang.reflect.Method.invoke (Method.java)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:494)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:964)
Caused by: java.lang.StringIndexOutOfBoundsException: 
  at java.lang.String.substring (String.java:2036)
  at in.basulabs.shakealarmclock.Activity_RingtonePicker.onActivityResult (Activity_RingtonePicker.java:100)
  at android.app.Activity.dispatchActivityResult (Activity.java:7759)
  at android.app.ActivityThread.deliverResults (ActivityThread.java:4610)
  at android.app.ActivityThread.handleSendResult (ActivityThread.java:4659)
  at android.app.servertransaction.ActivityResultItem.execute (ActivityResultItem.java:49)
  at android.app.servertransaction.TransactionExecutor.executeCallbacks (TransactionExecutor.java:108)
  at android.app.servertransaction.TransactionExecutor.execute (TransactionExecutor.java:68)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1955)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loop (Looper.java:214)
  at android.app.ActivityThread.main (ActivityThread.java:7078)
  at java.lang.reflect.Method.invoke (Method.java)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:494)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:964)
```

The exception _seems_ to be thrown from the following statement:

```java
String fileNameWithoutExt = fileNameWithExt.substring(0, fileNameWithExt.indexOf("."));
```

A file is being chosen by the user that does not have an extension?

--- a/.idea/compiler.xml
+++ b/.idea/compiler.xml
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
   <component name="CompilerConfiguration">
-    <bytecodeTargetLevel target="14" />
+    <bytecodeTargetLevel target="15" />
   </component>
 </project>

--- a/.idea/misc.xml
+++ b/.idea/misc.xml
@@ -51,7 +51,7 @@
       </value>
     </option>
   </component>
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_14_PREVIEW" default="false" project-jdk-name="14" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_X" default="true" project-jdk-name="15" project-jdk-type="JavaSDK">
     <output url="file://$PROJECT_DIR$/build/classes" />
   </component>
   <component name="ProjectType">

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -11,9 +11,16 @@ However, before contributing, please read the following guidelines:
    - Gradle version
    - Android SDK version
    - Build tools version
-   - Version of any other dependency, even if it is outdated
+   - Version of any other third-party library, even if it is outdated
    - App version and/or version codes
 1. Please try to make a logical series of commits in your PR such that they have meaning when someone looks at the history of the project. If this is not followed, I will do a squash merge rather than normal merge (doing so will put all your commits into one single commit in the history of your project).
 1. Clear merge conflicts, if any, that Github shows when you submit the PR. I will help if you need.
 1. Follow the "one feature one PR" rule: each PR should add only one feature to the app. You can change as many files as you need for introducing the feature, but these changes should cater to only one feature.
 1. Document your code where necessary. Changes without proper documentation may not be accepted.
+
+## Version information:
+The following versions have been used in compiling this project:
+
+1. Gradle version: 6.7.1
+1. Android Gradle Plugin version: 4.1.1
+1. JDK version: 15.0.1

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -8,8 +8,8 @@ android {
         applicationId "in.basulabs.shakealarmclock"
         minSdkVersion 21
         targetSdkVersion 30
-        versionCode 14
-        versionName "1.2.11"
+        versionCode 15
+        versionName "1.2.12"
 
         testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
     }

--- a/app/proguard-rules.pro
+++ b/app/proguard-rules.pro
@@ -14,7 +14,7 @@
 
 # Uncomment this to preserve the line number information for
 # debugging stack traces.
-#-keepattributes SourceFile,LineNumberTable
+-keepattributes SourceFile,LineNumberTable
 
 # If you keep the line number information, uncomment this to
 # hide the original source file name.

--- a/app/src/main/java/in/basulabs/shakealarmclock/Activity_AlarmsList.java
+++ b/app/src/main/java/in/basulabs/shakealarmclock/Activity_AlarmsList.java
@@ -12,6 +12,7 @@ import android.os.Bundle;
 import android.os.Handler;
 import android.os.Looper;
 import android.os.Message;
+import android.os.PowerManager;
 import android.view.Menu;
 import android.view.MenuItem;
 import android.view.View;
@@ -22,6 +23,7 @@ import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 import androidx.appcompat.app.AppCompatActivity;
 import androidx.appcompat.app.AppCompatDelegate;
+import androidx.fragment.app.DialogFragment;
 import androidx.lifecycle.ViewModelProvider;
 import androidx.recyclerview.widget.LinearLayoutManager;
 import androidx.recyclerview.widget.RecyclerView;
@@ -125,12 +127,11 @@ public class Activity_AlarmsList extends AppCompatActivity implements AlarmAdapt
 				}
 
 				startActivityForResult(intent, NEW_ALARM_REQUEST_CODE);
-
 			}
 		}
 
 		if (savedInstanceState == null && showAppUpdate) {
-			checkForUpdatesFromGooglePlay();
+			showDialogs();
 		}
 	}
 
@@ -345,7 +346,6 @@ public class Activity_AlarmsList extends AppCompatActivity implements AlarmAdapt
 						deleteOrDeactivateAlarm(MODE_DELETE_ALARM,
 								data.getInt(ConstantsAndStatics.BUNDLE_KEY_ALARM_HOUR),
 								data.getInt(ConstantsAndStatics.BUNDLE_KEY_ALARM_MINUTE));
-
 					}
 
 					AlarmEntity alarmEntity = new AlarmEntity(data.getInt(ConstantsAndStatics.BUNDLE_KEY_ALARM_HOUR),
@@ -471,6 +471,31 @@ public class Activity_AlarmsList extends AppCompatActivity implements AlarmAdapt
 
 	//-------------------------------------------------------------------------------------------------------------
 
+	private void showDialogs() {
+
+		boolean showBatteryOptimDialog = getSharedPreferences(ConstantsAndStatics.SHARED_PREF_FILE_NAME, MODE_PRIVATE)
+						.getBoolean(ConstantsAndStatics.SHARED_PREF_KEY_SHOW_BATTERY_OPTIM_DIALOG, true);
+
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && showBatteryOptimDialog) {
+			PowerManager powerManager = (PowerManager) getSystemService(POWER_SERVICE);
+			if (! powerManager.isIgnoringBatteryOptimizations("in.basulabs.shakealarmclock")) {
+				DialogFragment dialogFragment = new AlertDialog_BatteryOptimizations();
+				dialogFragment.setCancelable(false);
+				dialogFragment.show(getSupportFragmentManager(), "");
+				return;
+			}
+		}
+		checkForUpdatesFromGooglePlay();
+	}
+
+	//-------------------------------------------------------------------------------------------------------------
+
+	/**
+	 * Checks for updates from Google Play.
+	 * <p>
+	 *     Courtsey: https://github.com/javiersantos/AppUpdater
+	 * </p>
+	 */
 	private void checkForUpdatesFromGooglePlay() {
 		Context context = this;
 
@@ -494,6 +519,13 @@ public class Activity_AlarmsList extends AppCompatActivity implements AlarmAdapt
 
 	//-------------------------------------------------------------------------------------------------------------
 
+	/**
+	 * Checks for updates from Github.
+	 * <p>
+	 *     Courtsey: https://github.com/javiersantos/AppUpdater
+	 * </p>
+	 */
+	@SuppressWarnings({"unused", "RedundantSuppression"})
 	private void checkForUpdatesFromGitHub() {
 		Context context = this;
 

--- a/app/src/main/java/in/basulabs/shakealarmclock/Activity_RingtonePicker.java
+++ b/app/src/main/java/in/basulabs/shakealarmclock/Activity_RingtonePicker.java
@@ -164,7 +164,7 @@ public class Activity_RingtonePicker extends AppCompatActivity implements View.O
 				if (intent.hasExtra(EXTRA_RINGTONE_TITLE)) {
 					viewModel.setTitle((CharSequence) Objects.requireNonNull(intent.getExtras()).get(EXTRA_RINGTONE_TITLE));
 				} else {
-					viewModel.setTitle("Select tone:");
+					viewModel.setTitle((CharSequence) getResources().getString(R.string.ringtonePicker_defaultTitle));
 				}
 
 				if (intent.hasExtra(ConstantsAndStatics.EXTRA_PLAY_RINGTONE)) {
@@ -284,8 +284,7 @@ public class Activity_RingtonePicker extends AppCompatActivity implements View.O
 					// or not. If it exists, we shall add that file to our RadioGroup.
 					// If the file does not exist, we do not select any Radiogroup.
 					///////////////////////////////////////////////////////////////////////
-					try (Cursor cursor = getContentResolver()
-							.query(viewModel.getExistingUri(), null, null, null, null)) {
+					try (Cursor cursor = getContentResolver().query(viewModel.getExistingUri(), null, null, null, null)) {
 
 						if (cursor != null && cursor.getCount() > 0 && cursor.moveToFirst()) {
 							// existingUri is a valid Uri.
@@ -413,13 +412,15 @@ public class Activity_RingtonePicker extends AppCompatActivity implements View.O
 	 * Fires an implicit Intent to open a file browser and let the user choose an alarm tone.
 	 */
 	private void openFileBrowser() {
-		Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
-		intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
-		intent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
-		intent.addFlags(Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION);
-		intent.setType("*/*");
 		String[] mimeTypes = new String[]{"audio/mpeg", "audio/ogg", "audio/aac", "audio/x-matroska"};
-		intent.putExtra(Intent.EXTRA_MIME_TYPES, mimeTypes);
+
+		Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT)
+				.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
+				.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
+				.addFlags(Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION)
+				.setType("*/*")
+				.putExtra(Intent.EXTRA_MIME_TYPES, mimeTypes);
+
 		startActivityForResult(intent, FILE_REQUEST_CODE);
 	}
 
@@ -431,8 +432,7 @@ public class Activity_RingtonePicker extends AppCompatActivity implements View.O
 	 * @return {@code true} if the permission is available, otherwise {@code false}.
 	 */
 	private boolean isPermissionAvailable() {
-		return ContextCompat.checkSelfPermission(this,
-				Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED;
+		return ContextCompat.checkSelfPermission(this,	Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED;
 	}
 
 	//--------------------------------------------------------------------------------------------------
@@ -476,8 +476,7 @@ public class Activity_RingtonePicker extends AppCompatActivity implements View.O
 	 * Shows an {@code AlertDialog} explaining why the permission is necessary.
 	 */
 	private void showPermissionExplanationDialog() {
-		DialogFragment dialogPermissionReason = new AlertDialog_PermissionReason(
-				getResources().getString(R.string.permissionReasonExp_ringtonePicker));
+		DialogFragment dialogPermissionReason = new AlertDialog_PermissionReason(getResources().getString(R.string.permissionReasonExp_ringtonePicker));
 		dialogPermissionReason.setCancelable(false);
 		dialogPermissionReason.show(getSupportFragmentManager(), "");
 	}
@@ -520,15 +519,19 @@ public class Activity_RingtonePicker extends AppCompatActivity implements View.O
 							int nameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
 							cursor.moveToFirst();
 
-							String fileNameWithExt = cursor.getString(nameIndex);
-							String fileNameWithoutExt = fileNameWithExt.substring(0, fileNameWithExt.indexOf("."));
+							String fileName = cursor.getString(nameIndex);
+
+							int indexOfDot = fileName.lastIndexOf(".");
+							if (indexOfDot != -1){
+								fileName = fileName.substring(0, indexOfDot);
+							}
 							int toneId = View.generateViewId();
 
-							viewModel.getToneNameList().add(fileNameWithoutExt);
+							viewModel.getToneNameList().add(fileName);
 							viewModel.getToneUriList().add(toneUri);
 							viewModel.getToneIdList().add(toneId);
 
-							createOneRadioButton(toneId, fileNameWithoutExt);
+							createOneRadioButton(toneId, fileName);
 
 							((RadioButton) findViewById(toneId)).setChecked(true);
 						}
@@ -546,9 +549,8 @@ public class Activity_RingtonePicker extends AppCompatActivity implements View.O
 	public void onDialogPositiveClick(DialogFragment dialogFragment) {
 		if (dialogFragment.getClass().equals(AlertDialog_PermissionReason.class)) {
 
-			if ((! ActivityCompat.shouldShowRequestPermissionRationale(this,
-					Manifest.permission.READ_EXTERNAL_STORAGE)) && (sharedPreferences
-					.getBoolean(ConstantsAndStatics.SHARED_PREF_KEY_PERMISSION_WAS_ASKED_BEFORE, false))) {
+			if ((! ActivityCompat.shouldShowRequestPermissionRationale(this,	Manifest.permission.READ_EXTERNAL_STORAGE)) &&
+					(sharedPreferences.getBoolean(ConstantsAndStatics.SHARED_PREF_KEY_PERMISSION_WAS_ASKED_BEFORE, false))) {
 
 				////////////////////////////////////////////////////////////////////////////////
 				// User had chosen "Don't ask again". We need to redirect the user to the

--- a/app/src/main/java/in/basulabs/shakealarmclock/AlarmDatabase.java
+++ b/app/src/main/java/in/basulabs/shakealarmclock/AlarmDatabase.java
@@ -17,8 +17,7 @@ public abstract class AlarmDatabase extends RoomDatabase {
 
 	public static synchronized AlarmDatabase getInstance(Context context) {
 		if (instance == null) {
-			instance = Room.databaseBuilder(context.getApplicationContext(), AlarmDatabase.class,
-					DATABASE_NAME).fallbackToDestructiveMigration().build();
+			instance = Room.databaseBuilder(context.getApplicationContext(), AlarmDatabase.class, DATABASE_NAME).fallbackToDestructiveMigration().build();
 		}
 		return instance;
 	}

--- a/app/src/main/java/in/basulabs/shakealarmclock/AlertDialog_BatteryOptimizations.java
+++ b/app/src/main/java/in/basulabs/shakealarmclock/AlertDialog_BatteryOptimizations.java
@@ -1,0 +1,62 @@
+package in.basulabs.shakealarmclock;
+
+import android.app.Dialog;
+import android.content.Context;
+import android.content.Intent;
+import android.os.Bundle;
+import android.provider.Settings;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.appcompat.app.AlertDialog;
+import androidx.fragment.app.DialogFragment;
+
+public class AlertDialog_BatteryOptimizations extends DialogFragment {
+
+	private boolean attached;
+
+	@Override
+	public void onAttach(@NonNull Context context) {
+		super.onAttach(context);
+		attached = true;
+	}
+
+	//-------------------------------------------------------------------------------------------------------------
+
+	@NonNull
+	@Override
+	public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
+
+		AlertDialog.Builder builder = new AlertDialog.Builder(requireContext())
+				.setCancelable(false)
+				.setTitle(R.string.batteryOptim_title)
+				.setMessage(R.string.batteryOptim_message)
+				.setPositiveButton(R.string.batteryOptim_positive, (dialog, which) -> {
+					if (attached && android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
+						Intent intent = new Intent(Settings.ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS);
+						requireContext().startActivity(intent);
+					}
+
+				})
+				.setNegativeButton(R.string.batteryOptim_negative, null)
+				.setNeutralButton(R.string.batteryOptim_neutral, (dialog, which) -> {
+					dialog.dismiss();
+					requireContext().getSharedPreferences(ConstantsAndStatics.SHARED_PREF_FILE_NAME, Context.MODE_PRIVATE)
+							.edit()
+							.remove(ConstantsAndStatics.SHARED_PREF_KEY_SHOW_BATTERY_OPTIM_DIALOG)
+							.putBoolean(ConstantsAndStatics.SHARED_PREF_KEY_SHOW_BATTERY_OPTIM_DIALOG, false)
+							.commit();
+				});
+
+		return builder.create();
+	}
+
+	//--------------------------------------------------------------------------------------------------------------------
+
+	@Override
+	public void onDetach() {
+		super.onDetach();
+		attached = false;
+	}
+
+}

--- a/app/src/main/java/in/basulabs/shakealarmclock/ConstantsAndStatics.java
+++ b/app/src/main/java/in/basulabs/shakealarmclock/ConstantsAndStatics.java
@@ -29,58 +29,58 @@ final class ConstantsAndStatics {
 	 * Bundle key for the Bundle that is passed with intent from {@link Activity_AlarmDetails} to {@link Activity_AlarmsList} containing the data set by the
 	 * user.
 	 */
-	static final String BUNDLE_KEY_ALARM_DETAILS = "in.basulabs.shakealarmclock.BundleWithAlarmDetails";
+	static final String BUNDLE_KEY_ALARM_DETAILS = "in.basulabs.shakealarmclock.ALARM_DETAILS_BUNDLE";
 
 	/**
 	 * Bundle key for the alarm hour. The value is an integer.
 	 */
-	static final String BUNDLE_KEY_ALARM_HOUR = "in.basulabs.shakealarmclock.HourPickedByUser";
+	static final String BUNDLE_KEY_ALARM_HOUR = "in.basulabs.shakealarmclock.ALARM_HOUR";
 
 	/**
 	 * Bundle key for the alarm minute. The value is an integer.
 	 */
-	static final String BUNDLE_KEY_ALARM_MINUTE = "in.basulabs.shakealarmclock.MinsPickedByUser";
+	static final String BUNDLE_KEY_ALARM_MINUTE = "in.basulabs.shakealarmclock.ALARM_MINUTES";
 
 	/**
 	 * Bundle key for the alarm type. The value is one of {@link #ALARM_TYPE_SOUND_ONLY}, {@link #ALARM_TYPE_VIBRATE_ONLY} or {@link
 	 * #ALARM_TYPE_SOUND_AND_VIBRATE}.
 	 */
-	static final String BUNDLE_KEY_ALARM_TYPE = "in.basulabs.shakealarmclock.AlarmType";
+	static final String BUNDLE_KEY_ALARM_TYPE = "in.basulabs.shakealarmclock.ALARM_TYPE";
 
 	/**
 	 * Bundle key for the alarm volume. The value is an integer.
 	 */
-	static final String BUNDLE_KEY_ALARM_VOLUME = "in.basulabs.shakealarmclock.AlarmVolume";
+	static final String BUNDLE_KEY_ALARM_VOLUME = "in.basulabs.shakealarmclock.ALARM_VOLUME";
 
 	/**
 	 * Bundle key for the alarm snooze interval. The value is an integer.
 	 */
-	static final String BUNDLE_KEY_SNOOZE_TIME_IN_MINS = "in.basulabs.shakealarmclock.SnoozeTimeInMins";
+	static final String BUNDLE_KEY_SNOOZE_TIME_IN_MINS = "in.basulabs.shakealarmclock.SNOOZE_TIME_IN_MINS";
 
 	/**
 	 * Bundle key for the number of times the alarm should snooze itself. The value is an integer.
 	 */
-	static final String BUNDLE_KEY_SNOOZE_FREQUENCY = "in.basulabs.shakealarmclock.SnoozeFrequency";
+	static final String BUNDLE_KEY_SNOOZE_FREQUENCY = "in.basulabs.shakealarmclock.SNOOZE_FREQUENCY";
 
 	/**
 	 * Bundle key denoting whether repeat is on or off. Value is boolean.
 	 */
-	static final String BUNDLE_KEY_IS_REPEAT_ON = "in.basulabs.shakealarmclock.IsRepeatOn";
+	static final String BUNDLE_KEY_IS_REPEAT_ON = "in.basulabs.shakealarmclock.IS_REPEAT_ON";
 
 	/**
 	 * Bundle key denoting whether snooze is on or off. Value is boolean.
 	 */
-	static final String BUNDLE_KEY_IS_SNOOZE_ON = "in.basulabs.shakealarmclock.IsSnoozeOn";
+	static final String BUNDLE_KEY_IS_SNOOZE_ON = "in.basulabs.shakealarmclock.IS_SNOOZE_ON";
 
 	/**
 	 * Bundle key denoting whether alarm is on or off. Value is boolean.
 	 */
-	static final String BUNDLE_KEY_IS_ALARM_ON = "in.basulabs.shakealarmclock.IsAlarmOn";
+	static final String BUNDLE_KEY_IS_ALARM_ON = "in.basulabs.shakealarmclock.IS_ALARM_ON";
 
 	/**
 	 * Bundle key for the alarm repeat days. The value is an ArrayList of Integer type. Monday is 1 and Sunday is 7.
 	 */
-	static final String BUNDLE_KEY_REPEAT_DAYS = "in.basulabs.shakealarmclock.ArrayListOfRepeatDays";
+	static final String BUNDLE_KEY_REPEAT_DAYS = "in.basulabs.shakealarmclock.REPEAT_DAYS";
 
 	/**
 	 * Denotes that the alarm type will be "Sound".
@@ -130,12 +130,12 @@ final class ConstantsAndStatics {
 	/**
 	 * Intent action delivered to {@link android.content.BroadcastReceiver} in {@link Service_RingAlarm} instructing it to snooze the alarm.
 	 */
-	static final String ACTION_SNOOZE_ALARM = "in.basulabs.shakealarmclock.Service_RingAlarm -- SNOOZE_ALARM";
+	static final String ACTION_SNOOZE_ALARM = "in.basulabs.shakealarmclock.SNOOZE_ALARM";
 
 	/**
 	 * Intent action delivered to {@link android.content.BroadcastReceiver} in {@link Service_RingAlarm} instructing it to cancel the alarm.
 	 */
-	static final String ACTION_CANCEL_ALARM = "in.basulabs.shakealarmclock.Service_RingAlarm -- CANCEL_ALARM";
+	static final String ACTION_CANCEL_ALARM = "in.basulabs.shakealarmclock.CANCEL_ALARM";
 
 	/**
 	 * The name of the {@link android.content.SharedPreferences} file for this app.
@@ -152,6 +152,10 @@ final class ConstantsAndStatics {
 	 */
 	static final String ACTION_EXISTING_ALARM = "in.basulabs.shakealarmclock.ACTION_EXISTING_ALARM";
 
+	/**
+	 * Intent action indicating that a new alarm is being requested to be created from a direct intent to the app rather than the user clicking on the "Add"
+	 * button.
+	 */
 	static final String ACTION_NEW_ALARM_FROM_INTENT =	"in.basulabs.shakealarmclock.ACTION_NEW_ALARM_FROM_INTENT";
 
 	/**
@@ -208,8 +212,20 @@ final class ConstantsAndStatics {
 	 */
 	static final String SHARED_PREF_KEY_DEFAULT_POWER_BTN_OPERATION = "in.basulabs.shakealarmclock.DEFAULT_POWER_BTN_OPERATION";
 
+	/**
+	 * {@link android.content.SharedPreferences} key to store the sensitivity of the shake detector. The data type is {@code float}.
+	 */
 	static final String SHARED_PREF_KEY_SHAKE_SENSITIVITY = "in.basulabs.shakealarmclock.SHAKE_SENSITIVITY";
 
+	/**
+	 * {@link android.content.SharedPreferences} key indicating whether {@link AlertDialog_BatteryOptimizations} should be shown in
+	 * {@link Activity_AlarmsList}. Data type is {@code boolean}.
+	 */
+	static final String SHARED_PREF_KEY_SHOW_BATTERY_OPTIM_DIALOG = "in.basulabs.shakealarmclock.SHOW_BATTERY_OPTIM_DIALOG";
+
+	/**
+	 * The default sensitivity of the shake detector.
+	 */
 	static final float DEFAULT_SHAKE_SENSITIVITY = 3.2f;
 
 	/**
@@ -278,6 +294,10 @@ final class ConstantsAndStatics {
 	 */
 	static final String SHARED_PREF_KEY_THEME = "in.basulabs.shakealarmclock.THEME";
 
+	/**
+	 * {@link android.content.SharedPreferences} key indicating whether a new alarm tone chosen by the user should be set as the default tone for future
+	 * alarms. Data type: {@code boolean}.
+	 */
 	static final String SHARED_PREF_KEY_AUTO_SET_TONE = "in.basulabs.shakealarmclock.AUTO_SET_TONE";
 
 	/**

--- a/app/src/main/java/in/basulabs/shakealarmclock/Service_RingAlarm.java
+++ b/app/src/main/java/in/basulabs/shakealarmclock/Service_RingAlarm.java
@@ -25,6 +25,7 @@ import android.os.Bundle;
 import android.os.CountDownTimer;
 import android.os.Handler;
 import android.os.IBinder;
+import android.os.SystemClock;
 import android.os.VibrationEffect;
 import android.os.Vibrator;
 
@@ -506,6 +507,7 @@ public class Service_RingAlarm extends Service implements SensorEventListener {
 			} else {
 				vibrator.vibrate(200);
 			}
+			SystemClock.sleep(200);
 		}
 	}
 

--- a/app/src/main/java/in/basulabs/shakealarmclock/Service_SnoozeAlarm.java
+++ b/app/src/main/java/in/basulabs/shakealarmclock/Service_SnoozeAlarm.java
@@ -144,7 +144,7 @@ public class Service_SnoozeAlarm extends Service {
 		NotificationCompat.Builder builder = new NotificationCompat.Builder(this, Integer.toString(NOTIFICATION_ID))
 				.setContentTitle(getResources().getString(R.string.app_name))
 				.setContentText(getResources().getString(R.string.notifContent_snooze))
-				.setPriority(NotificationCompat.PRIORITY_MAX)
+				.setPriority(NotificationCompat.PRIORITY_HIGH)
 				.setCategory(NotificationCompat.CATEGORY_ALARM)
 				.setSmallIcon(R.drawable.ic_notif)
 				.setContentIntent(contentPendingIntent);

--- a/app/src/main/java/in/basulabs/shakealarmclock/Service_UpdateAlarm.java
+++ b/app/src/main/java/in/basulabs/shakealarmclock/Service_UpdateAlarm.java
@@ -7,20 +7,27 @@ import android.app.NotificationManager;
 import android.app.PendingIntent;
 import android.app.Service;
 import android.content.Intent;
+import android.content.pm.ServiceInfo;
 import android.os.Build;
 import android.os.Bundle;
 import android.os.IBinder;
+import android.text.format.DateFormat;
 
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 import androidx.core.app.NotificationCompat;
 
+import java.time.DayOfWeek;
 import java.time.LocalDate;
 import java.time.LocalDateTime;
 import java.time.LocalTime;
 import java.time.ZoneId;
 import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.TemporalAdjusters;
 import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Locale;
 import java.util.concurrent.atomic.AtomicReference;
 
 public class Service_UpdateAlarm extends Service {
@@ -36,7 +43,11 @@ public class Service_UpdateAlarm extends Service {
 	@Override
 	public int onStartCommand(Intent intent, int flags, int startId) {
 
-		startForeground(NOTIFICATION_ID, buildNotification());
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+			startForeground(NOTIFICATION_ID, buildNotification(), ServiceInfo.FOREGROUND_SERVICE_TYPE_NONE);
+		} else {
+			startForeground(NOTIFICATION_ID, buildNotification());
+		}
 		isThisServiceRunning = true;
 
 		ConstantsAndStatics.cancelScheduledPeriodicWork(this);
@@ -69,15 +80,16 @@ public class Service_UpdateAlarm extends Service {
 	//--------------------------------------------------------------------------------------------------
 
 	/**
-	 * Creates the notification channel.
+	 * Creates a notification channel.
+	 *
+	 * @param NOTIFICATION_ID The ID of the notification channel.
 	 */
-	private void createNotificationChannel() {
+	private void createNotificationChannel(final int NOTIFICATION_ID) {
 		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
 			int importance = NotificationManager.IMPORTANCE_HIGH;
 			NotificationChannel channel = new NotificationChannel(Integer.toString(NOTIFICATION_ID),
 					"in.basulabs.shakealarmclock Notification", importance);
 			NotificationManager notificationManager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
-			channel.setSound(null, null);
 			assert notificationManager != null;
 			notificationManager.createNotificationChannel(channel);
 		}
@@ -93,15 +105,14 @@ public class Service_UpdateAlarm extends Service {
 	 *        Notification)}.
 	 */
 	private Notification buildNotification() {
-		createNotificationChannel();
+		createNotificationChannel(NOTIFICATION_ID);
 
 		Intent contentIntent = new Intent(this, Activity_AlarmsList.class);
 		contentIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
 		contentIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK);
 		PendingIntent contentPendingIntent = PendingIntent.getActivity(this, 5701, contentIntent, PendingIntent.FLAG_UPDATE_CURRENT);
 
-		NotificationCompat.Builder builder = new NotificationCompat.Builder(this,
-				Integer.toString(NOTIFICATION_ID))
+		NotificationCompat.Builder builder = new NotificationCompat.Builder(this, Integer.toString(NOTIFICATION_ID))
 				.setContentTitle(getResources().getString(R.string.app_name))
 				.setContentText(getResources().getString(R.string.updateAlarm_notifMessage))
 				.setPriority(NotificationCompat.PRIORITY_DEFAULT)
@@ -166,6 +177,14 @@ public class Service_UpdateAlarm extends Service {
 
 	//--------------------------------------------------------------------------------------------------
 
+	/**
+	 * Get the list of repeat days for a particular alarm.
+	 *
+	 * @param alarmID The unique alarm ID.
+	 *
+	 * @return An {@link ArrayList} of integers for the days in which the alarm is repeated. The day numbers follow the {@link DayOfWeek} enum.
+	 */
+	@Nullable
 	private ArrayList<Integer> getRepeatDays(int alarmID) {
 
 		AtomicReference<ArrayList<Integer>> repeatDays = new AtomicReference<>();
@@ -183,6 +202,16 @@ public class Service_UpdateAlarm extends Service {
 
 	//--------------------------------------------------------------------------------------------------
 
+	/**
+	 * Activates the alarms that are switched on, if possibe.
+	 * <p>
+	 * If repeat is ON for an alarm, then the alarm is set as usual. If, however, repeat is OFF, then it is first checked whether the time is reachable or not.
+	 * If reachable, the alarm is set, otherwise the alarm is switched off in the database and a notification is posted using {@link
+	 * #postAlarmMissedNotification(int, LocalTime)}.
+	 * </p>
+	 *
+	 * @param alarmEntityArrayList The list of active alarms.
+	 */
 	private void activateAlarms(@NonNull ArrayList<AlarmEntity> alarmEntityArrayList) {
 
 		AlarmManager alarmManager = (AlarmManager) getSystemService(ALARM_SERVICE);
@@ -191,13 +220,14 @@ public class Service_UpdateAlarm extends Service {
 
 			ArrayList<Integer> repeatDays = getRepeatDays(alarmEntity.alarmID);
 
-			LocalDateTime alarmDateTime = ConstantsAndStatics.getAlarmDateTime(LocalDate.of(alarmEntity.alarmYear, alarmEntity.alarmMonth,
-					alarmEntity.alarmDay), LocalTime.of(alarmEntity.alarmHour, alarmEntity.alarmMinutes), alarmEntity.isRepeatOn, repeatDays);
+			LocalDateTime alarmDateTime;
 
-			/*LocalDate alarmDate = LocalDate.of(alarmEntity.alarmYear, alarmEntity.alarmMonth, alarmEntity.alarmDay);
+			LocalDate alarmDate = LocalDate.of(alarmEntity.alarmYear, alarmEntity.alarmMonth, alarmEntity.alarmDay);
 			LocalTime alarmTime = LocalTime.of(alarmEntity.alarmHour, alarmEntity.alarmMinutes);
 
 			if (alarmEntity.isRepeatOn && repeatDays != null && repeatDays.size() > 0) {
+
+				// If repeat is ON, set the alarm as we normally would.
 
 				Collections.sort(repeatDays);
 
@@ -215,42 +245,94 @@ public class Service_UpdateAlarm extends Service {
 						// There is a day available in the same week for the alarm to ring;
 						// select that day and break from loop.
 						////////////////////////////////////////////////////////////////////////
-						alarmDateTime =
-								alarmDateTime.with(TemporalAdjusters.next(DayOfWeek.of(repeatDays.get(i))));
+						alarmDateTime = alarmDateTime.with(TemporalAdjusters.next(DayOfWeek.of(repeatDays.get(i))));
 						break;
 					}
 					if (i == repeatDays.size() - 1) {
 						// No day possible in this week. Select the first available date from next week.
-						alarmDateTime = alarmDateTime
-								.with(TemporalAdjusters.next(DayOfWeek.of(repeatDays.get(0))));
+						alarmDateTime = alarmDateTime.with(TemporalAdjusters.next(DayOfWeek.of(repeatDays.get(0))));
 					}
 				}
 
 			} else {
 
+				// If repeat is OFF, first check whether the alarm time is reachable. If yes, then set the alarm, otherwise ignore this alarm and
+				// switch it off in the database.
+
 				alarmDateTime = LocalDateTime.of(alarmDate, alarmTime);
 
 				if (! alarmDateTime.isAfter(LocalDateTime.now())) {
-					alarmDateTime = alarmDateTime.plusDays(1);
+					alarmDateTime = null;
 				}
 
-			}*/
+			}
 
-			Intent intent = new Intent(getApplicationContext(), AlarmBroadcastReceiver.class);
-			intent.setAction(ConstantsAndStatics.ACTION_DELIVER_ALARM);
-			intent.setFlags(Intent.FLAG_RECEIVER_FOREGROUND);
+			if (alarmDateTime != null) {
 
-			Bundle data = alarmEntity.getAlarmDetailsInABundle();
-			data.putIntegerArrayList(ConstantsAndStatics.BUNDLE_KEY_REPEAT_DAYS, repeatDays);
-			intent.putExtra(ConstantsAndStatics.BUNDLE_KEY_ALARM_DETAILS, data);
+				Intent intent = new Intent(getApplicationContext(), AlarmBroadcastReceiver.class);
+				intent.setAction(ConstantsAndStatics.ACTION_DELIVER_ALARM);
+				intent.setFlags(Intent.FLAG_RECEIVER_FOREGROUND);
 
-			PendingIntent pendingIntent = PendingIntent.getBroadcast(getApplicationContext(), alarmEntity.alarmID, intent, 0);
+				Bundle data = alarmEntity.getAlarmDetailsInABundle();
+				data.putIntegerArrayList(ConstantsAndStatics.BUNDLE_KEY_REPEAT_DAYS, repeatDays);
+				intent.putExtra(ConstantsAndStatics.BUNDLE_KEY_ALARM_DETAILS, data);
 
-			ZonedDateTime zonedDateTime = ZonedDateTime.of(alarmDateTime, ZoneId.systemDefault());
+				PendingIntent pendingIntent = PendingIntent.getBroadcast(getApplicationContext(), alarmEntity.alarmID, intent, 0);
 
-			alarmManager.setAlarmClock(new AlarmManager.AlarmClockInfo(zonedDateTime.toEpochSecond() * 1000, pendingIntent), pendingIntent);
+				ZonedDateTime zonedDateTime = ZonedDateTime.of(alarmDateTime, ZoneId.systemDefault());
+
+				alarmManager.setAlarmClock(new AlarmManager.AlarmClockInfo(zonedDateTime.toEpochSecond() * 1000, pendingIntent), pendingIntent);
+
+			} else {
+
+				Thread thread = new Thread(() -> alarmDatabase.alarmDAO().toggleAlarm(alarmEntity.alarmID, 0));
+
+				thread.start();
+				try {
+					thread.join();
+				} catch (InterruptedException ignored) {
+				}
+
+				postAlarmMissedNotification(alarmEntity.alarmID, alarmTime);
+
+			}
 
 		}
+
+	}
+
+	//--------------------------------------------------------------------------------------------------
+
+	/**
+	 * Posts a notification informing the user that an alarm has been missed.
+	 *
+	 * @param alarmID The unique alarm ID — will be used as notification ID.
+	 * @param alarmTime The alarm time.
+	 */
+	private void postAlarmMissedNotification(int alarmID, LocalTime alarmTime) {
+
+		createNotificationChannel(alarmID);
+
+		NotificationManager notificationManager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
+
+		DateTimeFormatter formatter;
+		if (! DateFormat.is24HourFormat(this)) {
+			formatter = DateTimeFormatter.ofPattern("hh:mm a", Locale.getDefault());
+		} else {
+			formatter = DateTimeFormatter.ofPattern("HH:mm", Locale.getDefault());
+		}
+		
+		NotificationCompat.Builder builder = new NotificationCompat.Builder(this, Integer.toString(alarmID))
+				.setContentTitle(getResources().getString(R.string.updateAlarm_alarmMissedTitle))
+				.setContentText(getString(R.string.updateAlarm_alarmMissedText, alarmTime.format(formatter)))
+				.setSmallIcon(R.drawable.ic_notif)
+				.setPriority(NotificationCompat.PRIORITY_HIGH)
+				.setCategory(NotificationCompat.CATEGORY_ERROR)
+				.setDefaults(Notification.DEFAULT_SOUND)
+				.setAutoCancel(true)
+				.setOngoing(false);
+
+		notificationManager.notify(alarmID, builder.build());
 
 	}
 

--- a/app/src/main/java/in/basulabs/shakealarmclock/ViewModel_AlarmsList.java
+++ b/app/src/main/java/in/basulabs/shakealarmclock/ViewModel_AlarmsList.java
@@ -96,8 +96,7 @@ public class ViewModel_AlarmsList extends ViewModel {
 						// Update the date iff the alarm is OFF and the repeat is OFF.
 						if (! entity.isRepeatOn && ! entity.isAlarmOn) {
 
-							alarmDateTime = LocalDateTime.of(entity.alarmYear, entity.alarmMonth,
-									entity.alarmDay, entity.alarmHour, entity.alarmMinutes);
+							alarmDateTime = LocalDateTime.of(entity.alarmYear, entity.alarmMonth, entity.alarmDay, entity.alarmHour, entity.alarmMinutes);
 
 							if (alarmDateTime.isBefore(LocalDateTime.now())) {
 								while (alarmDateTime.isBefore(LocalDateTime.now())) {
@@ -121,15 +120,11 @@ public class ViewModel_AlarmsList extends ViewModel {
 
 					for (AlarmEntity entity : alarmEntityList) {
 
-						LocalDateTime alarmDateTime = LocalDateTime.of(entity.alarmYear, entity.alarmMonth,
-								entity.alarmDay, entity.alarmHour, entity.alarmMinutes);
+						LocalDateTime alarmDateTime = LocalDateTime.of(entity.alarmYear, entity.alarmMonth,	entity.alarmDay, entity.alarmHour, entity.alarmMinutes);
 
-						ArrayList<Integer> repeatDays = entity.isRepeatOn ?
-								new ArrayList<>(
-										alarmDatabase.alarmDAO().getAlarmRepeatDays(entity.alarmID)) : null;
+						ArrayList<Integer> repeatDays = entity.isRepeatOn ?	new ArrayList<>(alarmDatabase.alarmDAO().getAlarmRepeatDays(entity.alarmID)) : null;
 
-						Objects.requireNonNull(alarmDataArrayList.getValue())
-								.add(getAlarmDataObject(entity, alarmDateTime, repeatDays));
+						Objects.requireNonNull(alarmDataArrayList.getValue()).add(getAlarmDataObject(entity, alarmDateTime, repeatDays));
 
 					}
 				}
@@ -198,10 +193,7 @@ public class ViewModel_AlarmsList extends ViewModel {
 	 * @return An array consiting of TWO elements: the alarm ID at index 0 and the position at which the alarm was inserted at index 1. The latter can be used
 	 * 		to scroll the {@link androidx.recyclerview.widget.RecyclerView} using {@link androidx.recyclerview.widget.RecyclerView#scrollToPosition(int)}.
 	 */
-	public int[] addAlarm(@NonNull AlarmDatabase alarmDatabase, @NonNull AlarmEntity alarmEntity,
-	                      @Nullable ArrayList<Integer> repeatDays) {
-
-		//Log.e(this.getClass().getSimpleName(), "addAlarm(...) called.");
+	public int[] addAlarm(@NonNull AlarmDatabase alarmDatabase, @NonNull AlarmEntity alarmEntity, @Nullable ArrayList<Integer> repeatDays) {
 
 		AtomicInteger alarmID = new AtomicInteger();
 
@@ -227,13 +219,10 @@ public class ViewModel_AlarmsList extends ViewModel {
 		///////////////////////////////////////////////////////////////
 		// Now, change the data in alarmDataArrayList.
 		///////////////////////////////////////////////////////////////
-		LocalDateTime alarmDateTime = LocalDateTime.of(alarmEntity.alarmYear, alarmEntity.alarmMonth,
-				alarmEntity.alarmDay, alarmEntity.alarmHour, alarmEntity.alarmMinutes);
+		LocalDateTime alarmDateTime = LocalDateTime.of(alarmEntity.alarmYear, alarmEntity.alarmMonth, alarmEntity.alarmDay, alarmEntity.alarmHour,
+				alarmEntity.alarmMinutes);
 
 		int scrollToPosition = 0;
-
-		/*Log.e(this.getClass().getSimpleName(), "size = " + Objects
-				.requireNonNull(alarmDataArrayList.getValue()).size());*/
 
 		AlarmData newAlarmData = getAlarmDataObject(alarmEntity, alarmDateTime, repeatDays);
 
@@ -242,42 +231,25 @@ public class ViewModel_AlarmsList extends ViewModel {
 			alarmDataArrayList = new MutableLiveData<>(new ArrayList<>());
 			Objects.requireNonNull(alarmDataArrayList.getValue()).add(newAlarmData);
 
-			//Log.e(this.getClass().getSimpleName(), "null");
-
 		} else {
-
-			//Log.e(this.getClass().getSimpleName(), "NOT null");
 
 			int index = isAlarmInTheList(alarmEntity.alarmHour, alarmEntity.alarmMinutes);
 			if (index != - 1) {
-				//Log.e(this.getClass().getSimpleName(), "alarm already in the list.");
 				alarmDataArrayList.getValue().remove(index);
 			}
 
-
 			for (int i = 0; i < Objects.requireNonNull(alarmDataArrayList.getValue()).size(); i++) {
 
-				//Log.e(this.getClass().getSimpleName(), "Inside loop.");
-
-				if (alarmDataArrayList.getValue().get(i).getAlarmDateTime().toLocalTime()
-						.isBefore(alarmDateTime.toLocalTime())) {
-
-					//Log.e(this.getClass().getSimpleName(), "Initial 1.");
+				if (alarmDataArrayList.getValue().get(i).getAlarmDateTime().toLocalTime().isBefore(alarmDateTime.toLocalTime())) {
 
 					if ((i + 1) < alarmDataArrayList.getValue().size()) {
 
-						//Log.e(this.getClass().getSimpleName(), "Initial 2.");
-
-						if (alarmDataArrayList.getValue().get(i + 1).getAlarmDateTime().toLocalTime()
-								.isAfter(alarmDateTime.toLocalTime())) {
-
-							//Log.e(this.getClass().getSimpleName(), "Initial 3.");
+						if (alarmDataArrayList.getValue().get(i + 1).getAlarmDateTime().toLocalTime().isAfter(alarmDateTime.toLocalTime())) {
 							alarmDataArrayList.getValue().add(i + 1, newAlarmData);
 							scrollToPosition = i + 1;
 							break;
 						}
 					} else {
-						//Log.e(this.getClass().getSimpleName(), "Initial 4.");
 						alarmDataArrayList.getValue().add(newAlarmData);
 						scrollToPosition = alarmDataArrayList.getValue().size() - 1;
 						break;
@@ -285,15 +257,12 @@ public class ViewModel_AlarmsList extends ViewModel {
 				}
 
 				if (i == alarmDataArrayList.getValue().size() - 1) {
-					//Log.e(this.getClass().getSimpleName(), "Initial 5.");
 					alarmDataArrayList.getValue().add(0, newAlarmData);
 					break;
 				}
 
 			}
 		}
-
-		//Log.e(this.getClass().getSimpleName(), "size = " + alarmDataArrayList.getValue().size());
 
 		try {
 			thread.join();
@@ -316,8 +285,7 @@ public class ViewModel_AlarmsList extends ViewModel {
 	 *
 	 * @return An {@link AlarmData} object that can be added to {@code alarmDataArrayList}.
 	 */
-	private AlarmData getAlarmDataObject(@NonNull AlarmEntity entity, @NonNull LocalDateTime alarmDateTime,
-	                                     @Nullable ArrayList<Integer> repeatDays) {
+	private AlarmData getAlarmDataObject(@NonNull AlarmEntity entity, @NonNull LocalDateTime alarmDateTime, @Nullable ArrayList<Integer> repeatDays) {
 
 		if (! entity.isRepeatOn) {
 			return new AlarmData(entity.isAlarmOn, alarmDateTime, entity.alarmType);
@@ -344,7 +312,6 @@ public class ViewModel_AlarmsList extends ViewModel {
 
 		Thread thread = new Thread(() -> {
 			alarmId.set(alarmDatabase.alarmDAO().getAlarmId(hour, mins));
-
 			alarmDatabase.alarmDAO().deleteAlarm(hour, mins);
 		});
 		thread.start();
@@ -441,9 +408,7 @@ public class ViewModel_AlarmsList extends ViewModel {
 	public ArrayList<Integer> getRepeatDays(@NonNull AlarmDatabase alarmDatabase, int hour, int mins) {
 		AtomicReference<ArrayList<Integer>> repeatDays = new AtomicReference<>(new ArrayList<>());
 
-		Thread thread = new Thread(() -> repeatDays
-				.set(new ArrayList<>(alarmDatabase.alarmDAO().getAlarmRepeatDays(getAlarmId(alarmDatabase,
-						hour, mins)))));
+		Thread thread = new Thread(() -> repeatDays.set(new ArrayList<>(alarmDatabase.alarmDAO().getAlarmRepeatDays(getAlarmId(alarmDatabase, hour, mins)))));
 		thread.start();
 
 		try {
@@ -468,8 +433,7 @@ public class ViewModel_AlarmsList extends ViewModel {
 	public AlarmEntity getAlarmEntity(@NonNull AlarmDatabase alarmDatabase, int hour, int mins) {
 		AtomicReference<AlarmEntity> alarmEntity = new AtomicReference<>();
 
-		Thread thread = new Thread(
-				() -> alarmEntity.set(alarmDatabase.alarmDAO().getAlarmDetails(hour, mins).get(0)));
+		Thread thread = new Thread(() -> alarmEntity.set(alarmDatabase.alarmDAO().getAlarmDetails(hour, mins).get(0)));
 		thread.start();
 
 		try {

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -87,6 +87,7 @@
     <string name="defaultTone">Default</string>
     <string name="silentTone">Silent</string>
     <string name="play_tone">Play tone</string>
+    <string name="ringtonePicker_defaultTitle">Select tone:</string>
 
     <!-- AlertDialog_DiscardChanges -->
     <string name="cancelDialogMessage">Discard changes?</string>

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -147,6 +147,14 @@
     <string name="message_shakeDetectionTest">Shake your phone. A vibration will tell you that a shake event has
     been detected. Change the sensitivity as per your choice.</string>
     <string name="negative_shakeDetectionTest">Stop</string>
+    
+    <!-- AlertDialog_BatteryOptimizations -->
+    <string name="batteryOptim_message">Please disable battery optimizations for this app. If this app is 
+        optimized, it might not be able to ring the alarm.</string>
+    <string name="batteryOptim_title">Disable battery optimizations</string>
+    <string name="batteryOptim_positive">Proceed</string>
+    <string name="batteryOptim_negative">Maybe later</string>
+    <string name="batteryOptim_neutral">Never show again</string>
 
 
     <string name="defaultAlarmToneText">Default alarm tone</string>

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -141,6 +141,8 @@
 
     <!-- Service_UpdateAlarm -->
     <string name="updateAlarm_notifMessage">Activating alarms. This may take a while…</string>
+    <string name="updateAlarm_alarmMissedTitle">Alarm missed!</string>
+    <string name="updateAlarm_alarmMissedText">The alarm at %s did not ring.</string>
 
     <!-- AlertDialog_TestShakeSensitivity -->
     <string name="title_shakeDetectionTest">Test sensitivity of shake detector</string>

--- a/build.gradle
+++ b/build.gradle
@@ -4,7 +4,7 @@ buildscript {
         jcenter()
     }
     dependencies {
-        classpath "com.android.tools.build:gradle:4.1.0"
+        classpath 'com.android.tools.build:gradle:4.1.1'
     }
 }
 

--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -1,6 +1,6 @@
-#Tue Jul 14 14:27:48 IST 2020
+#Sat Nov 21 11:14:36 IST 2020
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.5-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.7.1-bin.zip

